{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "31f2be86-6da9-4aec-91c6-10795affaaff",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "载入需要的程辑包：SingleCellExperiment\n",
      "\n",
      "载入需要的程辑包：SummarizedExperiment\n",
      "\n",
      "载入需要的程辑包：MatrixGenerics\n",
      "\n",
      "载入需要的程辑包：matrixStats\n",
      "\n",
      "\n",
      "载入程辑包：‘MatrixGenerics’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    colAlls, colAnyNAs, colAnys, colAvgsPerRowSet, colCollapse,\n",
      "    colCounts, colCummaxs, colCummins, colCumprods, colCumsums,\n",
      "    colDiffs, colIQRDiffs, colIQRs, colLogSumExps, colMadDiffs,\n",
      "    colMads, colMaxs, colMeans2, colMedians, colMins, colOrderStats,\n",
      "    colProds, colQuantiles, colRanges, colRanks, colSdDiffs, colSds,\n",
      "    colSums2, colTabulates, colVarDiffs, colVars, colWeightedMads,\n",
      "    colWeightedMeans, colWeightedMedians, colWeightedSds,\n",
      "    colWeightedVars, rowAlls, rowAnyNAs, rowAnys, rowAvgsPerColSet,\n",
      "    rowCollapse, rowCounts, rowCummaxs, rowCummins, rowCumprods,\n",
      "    rowCumsums, rowDiffs, rowIQRDiffs, rowIQRs, rowLogSumExps,\n",
      "    rowMadDiffs, rowMads, rowMaxs, rowMeans2, rowMedians, rowMins,\n",
      "    rowOrderStats, rowProds, rowQuantiles, rowRanges, rowRanks,\n",
      "    rowSdDiffs, rowSds, rowSums2, rowTabulates, rowVarDiffs, rowVars,\n",
      "    rowWeightedMads, rowWeightedMeans, rowWeightedMedians,\n",
      "    rowWeightedSds, rowWeightedVars\n",
      "\n",
      "\n",
      "载入需要的程辑包：GenomicRanges\n",
      "\n",
      "载入需要的程辑包：stats4\n",
      "\n",
      "载入需要的程辑包：BiocGenerics\n",
      "\n",
      "\n",
      "载入程辑包：‘BiocGenerics’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    IQR, mad, sd, var, xtabs\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    anyDuplicated, aperm, append, as.data.frame, basename, cbind,\n",
      "    colnames, dirname, do.call, duplicated, eval, evalq, Filter, Find,\n",
      "    get, grep, grepl, intersect, is.unsorted, lapply, Map, mapply,\n",
      "    match, mget, order, paste, pmax, pmax.int, pmin, pmin.int,\n",
      "    Position, rank, rbind, Reduce, rownames, sapply, setdiff, sort,\n",
      "    table, tapply, union, unique, unsplit, which.max, which.min\n",
      "\n",
      "\n",
      "载入需要的程辑包：S4Vectors\n",
      "\n",
      "\n",
      "载入程辑包：‘S4Vectors’\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    first, second\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    expand.grid, I, unname\n",
      "\n",
      "\n",
      "载入需要的程辑包：IRanges\n",
      "\n",
      "\n",
      "载入程辑包：‘IRanges’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:sp’:\n",
      "\n",
      "    %over%\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:data.table’:\n",
      "\n",
      "    shift\n",
      "\n",
      "\n",
      "载入需要的程辑包：GenomeInfoDb\n",
      "\n",
      "载入需要的程辑包：Biobase\n",
      "\n",
      "Welcome to Bioconductor\n",
      "\n",
      "    Vignettes contain introductory material; view with\n",
      "    'browseVignettes()'. To cite Bioconductor, see\n",
      "    'citation(\"Biobase\")', and for packages 'citation(\"pkgname\")'.\n",
      "\n",
      "\n",
      "\n",
      "载入程辑包：‘Biobase’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MatrixGenerics’:\n",
      "\n",
      "    rowMedians\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:matrixStats’:\n",
      "\n",
      "    anyMissing, rowMedians\n",
      "\n",
      "\n",
      "\n",
      "载入程辑包：‘SummarizedExperiment’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Seurat’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:SeuratObject’:\n",
      "\n",
      "    Assays\n",
      "\n",
      "\n",
      "载入需要的程辑包：scuttle\n",
      "\n",
      "载入需要的程辑包：ggplot2\n",
      "\n",
      "Package 'mclust' version 6.1\n",
      "Type 'citation(\"mclust\")' for citing this R package in publications.\n",
      "\n",
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 1.8 GiB”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n"
     ]
    }
   ],
   "source": [
    "#load the necessary packages\n",
    "library(data.table)\n",
    "library(Seurat, lib='/share/home/chenjy/SDcon/Seurat5.0.1')\n",
    "source('../R/main.R')\n",
    "source('../R/help.R')\n",
    "\n",
    "#load the datasets\n",
    "load(\"../scdata/MOB/MOB.dge.sceset.RData\")\n",
    "load(\"../spdata/MOB/Rep12_MOB_count_matrix-1.RData\")\n",
    "load(\"../spdata/MOB/Rep12_MOB_manual_annotation.RData\")\n",
    "\n",
    "count_matrix=as.data.frame(t(assays(sce)$counts))\n",
    "meta = as.data.frame(colData(sce))\n",
    "meta['cellType'][meta['cellType']=='M/TC']='M_TC'\n",
    "rownames(count_matrix)=paste0('cell', 1:nrow(count_matrix))\n",
    "rownames(meta)=paste0('cell', 1:nrow(meta))\n",
    "\n",
    "#construct the SeuratObject for scdata\n",
    "seu <- CreateSeuratObject(counts = t(count_matrix))\n",
    "active_ident=as.matrix(meta[,'cellType'])\n",
    "colnames(active_ident)=NULL\n",
    "rownames(active_ident)= rownames(meta)\n",
    "Idents(seu)=active_ident[,1]\n",
    "cell_names=rownames(as.matrix(Idents(seu)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "5ef53b5d-a8db-466a-b24a-18868ed7707b",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "#Quality control of the spatial transcriptomic data and corresponding spatial spots\n",
    "celltype=c('GC','PGC','M_TC','OSNs','EPL-IN')\n",
    "Num_celltype=length(celltype)\n",
    "\n",
    "geneexpr=as.matrix(t(MOB_raw))\n",
    "rownames(geneexpr) = paste0(\"spot\",1:nrow(geneexpr))\n",
    "threshold <- 20\n",
    "filtered_rows <- geneexpr[rowSums(geneexpr) > threshold, ]\n",
    "filtered_matrix <- filtered_rows[, colSums(geneexpr) > threshold]\n",
    "\n",
    "pos <- cbind.data.frame(x=as.numeric(sapply(strsplit(colnames(MOB_raw),split=\"x\"),\"[\",1)),y=as.numeric(sapply(strsplit(colnames(MOB_raw),split=\"x\"),\"[\",2)))\n",
    "colnames(pos) = c(\"x\", \"y\")\n",
    "rownames(pos) = paste0(\"spot\",1:nrow(pos))\n",
    "aligned_pos=align_pos(pos,filtered_matrix)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "f26f8939-f183-40d0-a516-3eadc6216222",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in asMethod(object):\n",
      "“sparse->dense coercion: allocating vector of size 1.8 GiB”\n",
      "Warning message:\n",
      "“Data is of class matrix. Coercing to dgCMatrix.”\n",
      "Running SCTransform on assay: RNA\n",
      "\n",
      "Running SCTransform on layer: counts\n",
      "\n",
      "vst.flavor='v2' set. Using model with fixed slope and excluding poisson genes.\n",
      "\n",
      "Variance stabilizing transformation of count matrix of size 15939 by 12801\n",
      "\n",
      "Model formula is y ~ log_umi\n",
      "\n",
      "Get Negative Binomial regression parameters per gene\n",
      "\n",
      "Using 2000 genes, 5000 cells\n",
      "\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Warning message:\n",
      "“useNames = NA is deprecated. Instead, specify either useNames = TRUE or useNames = FALSE.”\n",
      "Second step: Get residuals using fitted parameters for 15939 genes\n",
      "\n",
      "Computing corrected count matrix for 15939 genes\n",
      "\n",
      "Calculating gene attributes\n",
      "\n",
      "Wall clock passed: Time difference of 1.980398 mins\n",
      "\n",
      "Determine variable features\n",
      "\n",
      "Centering data matrix\n",
      "\n",
      "Getting residuals for block 1(of 3) for counts dataset\n",
      "\n",
      "Getting residuals for block 2(of 3) for counts dataset\n",
      "\n",
      "Getting residuals for block 3(of 3) for counts dataset\n",
      "\n",
      "Centering data matrix\n",
      "\n",
      "Finished calculating residuals for counts\n",
      "\n",
      "Set default assay to SCT\n",
      "\n",
      "Calculating cluster GC\n",
      "\n",
      "Calculating cluster PGC\n",
      "\n",
      "Calculating cluster M_TC\n",
      "\n",
      "Calculating cluster OSNs\n",
      "\n",
      "Calculating cluster EPL-IN\n",
      "\n",
      "\n",
      "载入程辑包：‘dplyr’\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:Biobase’:\n",
      "\n",
      "    combine\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:GenomicRanges’:\n",
      "\n",
      "    intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:GenomeInfoDb’:\n",
      "\n",
      "    intersect\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:IRanges’:\n",
      "\n",
      "    collapse, desc, intersect, setdiff, slice, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:S4Vectors’:\n",
      "\n",
      "    first, intersect, rename, setdiff, setequal, union\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:BiocGenerics’:\n",
      "\n",
      "    combine, intersect, setdiff, union\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:matrixStats’:\n",
      "\n",
      "    count\n",
      "\n",
      "\n",
      "The following object is masked from ‘package:MASS’:\n",
      "\n",
      "    select\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:data.table’:\n",
      "\n",
      "    between, first, last\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:stats’:\n",
      "\n",
      "    filter, lag\n",
      "\n",
      "\n",
      "The following objects are masked from ‘package:base’:\n",
      "\n",
      "    intersect, setdiff, setequal, union\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#Run DEG analysis on scdata to identify top 50 marker genes for each cell types\n",
    "\n",
    "selected_sc_count = as.matrix(seu@ assays$RNA$counts)\n",
    "selected_sc_meta =as.data.frame(Idents(seu))\n",
    "\n",
    "colnames(selected_sc_meta) = \"cellType\"\n",
    "selected_sc_meta$sampleInfo = rownames(selected_sc_meta)\n",
    "\n",
    "selection_seu <- CreateSeuratObject(counts = selected_sc_count,min.features = 200,min.cells=3, project = \"simulation\")\n",
    "\n",
    "active_ident=as.matrix(selected_sc_meta[,'cellType'])\n",
    "colnames(active_ident)=NULL\n",
    "rownames(active_ident)= rownames(selected_sc_meta)\n",
    "\n",
    "Idents(selection_seu)=active_ident[,1]\n",
    "cell_names=rownames(as.matrix(Idents(selection_seu)))\n",
    "\n",
    "selection_seu = SCTransform(selection_seu)\n",
    "all.markers <- FindAllMarkers(selection_seu)\n",
    "\n",
    "library(dplyr)\n",
    "all.markers %>%\n",
    "  group_by(cluster) %>%\n",
    "  top_n(n = 50, wt = avg_log2FC) -> top5\n",
    "\n",
    "top_DEGs =dplyr::rename(top5,'label'='cluster')\n",
    "top_DEGs=top_DEGs %>% arrange(label, desc(avg_log2FC))  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "9a82eb5f-87dd-4eac-9138-f2c62d6094bc",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "\n",
    "top_DEGs=top5 %>% arrange(label, desc(avg_log2FC))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "972e62a3-f507-40c1-ba8d-3db93aae8f6d",
   "metadata": {
    "tags": []
   },
   "outputs": [
    {
     "ename": "ERROR",
     "evalue": "Error in QR_SIDE(sp_expr, sp_pos, top_DEGs, Num_topic, Num_HVG, dim_embed, : 没有\"QR_SIDE\"这个函数\n",
     "output_type": "error",
     "traceback": [
      "Error in QR_SIDE(sp_expr, sp_pos, top_DEGs, Num_topic, Num_HVG, dim_embed, : 没有\"QR_SIDE\"这个函数\nTraceback:\n"
     ]
    }
   ],
   "source": [
    "#cell-type deconvolution using QR-SIDE\n",
    "source('../R/main.R')\n",
    "\n",
    "sp_expr=filtered_matrix # Spot by gene SRT matrix\n",
    "\n",
    "sp_pos=aligned_pos #Spot by 2, 2D spatial coordination\n",
    "\n",
    "top_DEGs=top_DEGs #DEGs dataframe from Seurat sorted by 'avg_log2FC'\n",
    "\n",
    "Num_topic=7 #Num of topic domains\n",
    "\n",
    "Num_HVG=1000 #num of highly variable genes used in QR_SIDE\n",
    "\n",
    "dim_embed=20 #dimension of embeddings in hierachical factor models\n",
    "\n",
    "top_marker_num=4 #Num of markers of each cell type from \"top_DEGs\" used in QR_SIDE \n",
    "\n",
    "fixed_marker_list=FALSE #whether QR-SIDE uses the all genes in \"top_DEGs\" as markers. If FALSE, \"top_marker_num\" must be provided\n",
    "\n",
    "out=QR_SIDE(sp_expr, sp_pos, top_DEGs, Num_topic, Num_HVG, dim_embed, top_marker_num, fixed_marker_list)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8d5259ad-7823-4389-bfc4-429df528f46f",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.2.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
